File: 2019/day6.go
package main

import (
	"github.com/lukechampine/advent/utils"
)

var input = utils.ReadInput("day6_input.txt")
var ils = utils.Lines(input)

func recOrbits(o string, orbits map[string]string) int {
	if _, ok := orbits[o]; !ok {
		return 0
	}
	return 1 + recOrbits(orbits[o], orbits)
}

func orbitPath(o string, orbits map[string]string) []string {
	if _, ok := orbits[o]; !ok {
		return nil
	}
	return append(orbitPath(orbits[o], orbits), o)
}

func main() {
	// part 1
	orbits := make(map[string]string)
	for _, line := range ils {
		xs := utils.Split(line, ")")
		orbits[xs[1]] = xs[0]
	}
	var n int
	for o := range orbits {
		n += recOrbits(o, orbits)
	}
	utils.Println(n)

	// part 2
	youPath := orbitPath("YOU", orbits)
	sanPath := orbitPath("SAN", orbits)

	for i := range youPath {
		if youPath[i] != sanPath[i] {
			utils.Println(len(youPath[i+1:]) + len(sanPath[i+1:]))
			return
		}
	}
}

File: 2019/day6.zig
const std = @import("std");
const utils = @import("utils.zig");

fn recOrbits(o: []const u8, orbits: var) usize {
    return if (orbits.getValue(o)) |parent|
        1 + recOrbits(parent, orbits)
    else
        0;
}

fn orbitPath(o: []const u8, orbits: var) [][]const u8 {
    return if (orbits.getValue(o)) |parent|
        utils.append([]const u8, orbitPath(parent, orbits), o)
    else
        utils.alloc([]const u8, 0);
}

pub fn main() void {
    const input = utils.readFileLines("day6_input.txt");

    var orbits = std.StringHashMap([]const u8).init(std.heap.direct_allocator);
    for (input) |line| {
        var xs = utils.splitByte(line, ')');
        _ = orbits.put(xs[1], xs[0]) catch unreachable;
    }

    var n: usize = 0;
    var it = orbits.iterator();
    while (it.next()) |o| {
        n += recOrbits(o.key, orbits);
    }
    utils.println(n);

    var youPath = orbitPath("YOU", orbits);
    var sanPath = orbitPath("SAN", orbits);

    for (youPath) |_, i| {
        if (!std.mem.eql(u8, youPath[i], sanPath[i])) {
            utils.println(youPath[i + 1 ..].len + sanPath[i + 1 ..].len);
            return;
        }
    }
}
